learn_evm/arithmetic-checks.md

```diff
@@ -0,0 +1,668 @@
+However, in this case, we don't save any operations.
+
+Just as with addition, for signed integers, we can combine the checks for both scenarios into a single check using `xor`.
+
+```solidity
+function checkedSubInt(int256 a, int256 b) public pure returns (int256 c) {
+    unchecked {
+        c = a - b;
+
+        bool overflow;
+
+        assembly {
+            // If `b >= 0`, then the result `c = a - b` can't be greater than `a`.
+            // If `b <  0`, then the result `c = a - b` can't be less than `a`.
+            overflow := xor(sgt(b, 0), sgt(a, c))
+        }
+
+        if (overflow) arithmeticError();
+    }
+}
+```
+
+## Arithmetic checks for uint256 multiplication
+
+To detect overflow when multiplying two unsigned integers, we can again use the approach of computing the maximum possible value of a multiplicand and check that it isn't exceeded.
+
+```solidity
+/// @notice versions >=0.8.0 && <0.8.17
+function checkedMulUint1(uint256 a, uint256 b) public pure returns (uint256 c) {
+    unchecked {
+        c = a * b;
+
+        if (a != 0 && b > type(uint256).max / a) arithmeticError();
+    }
+}
+```
+
+> It's important to note that the Solidity compiler always includes a division by zero check for all division and modulo operations, regardless of the presence of an unchecked block.
+> The EVM itself simply returns `0` when dividing by `0`, and this also applies to inline-assembly.
+> If the order of the boolean expressions is evaluated in reverse order, it could cause an arithmetic check to incorrectly revert when `a = 0`.
+
+We can compute the maximum value for `b` as long as `a` is non-zero. However, if `a` is zero, we know that the result will be zero as well, and there is no need to check for overflow.
+Like before, we can also make use of the result and try to reconstruct one multiplicand from it. This is possible if the product didn't overflow and the first multiplicand is non-zero.
+
+```solidity
+/// @notice versions >=0.8.17
+function checkedMulUint2(uint256 a, uint256 b) public pure returns (uint256 c) {
+    unchecked {
+        c = a * b;
+
+        if (a != 0 && b != c / a) arithmeticError();
+    }
+}
+```
+
+For reference, we can further remove the additional division by zero check by writing the code in assembly.
+
+```solidity
+function checkedMulUint3(uint256 a, uint256 b) public pure returns (uint256 c) {
+    unchecked {
+        c = a * b;
+
+        bool overflow;
+
+        assembly {
+            // This version does not include a redundant division-by-0 check
+            // which the Solidity compiler includes when performing `c / a`.
+            overflow := iszero(or(iszero(a), eq(div(c, a), b)))
+        }
+
+        if (overflow) arithmeticError();
+    }
+}
+```
+
+## Arithmetic checks for int256 multiplication
+
+In older versions, the Solidity compiler uses four separate checks to detect integer multiplication overflow.
```

Comment:
what are the older versions? <0.8.17?

Reply:
I included the exact versions in the code snippet, didn't want to clutter the text.